### PR TITLE
Created timestamp on arrival of log in queue (preventing Out of Order)

### LINKED ISSUE #49: Stops send log after Loki Error 'entry out of order' (Http status 400)
When our application has some load (a lot of broker messages to consume - using [Rebus](https://github.com/rebus-org/Rebus) worker threads), the sink suddenly stops sending logs to Loki.

To investigate, we configured the application to have 2 sinks enabled: console and this sink, and Promtail configured to send console log to Loki.
We found this line on Loki log:

`
level=warn ts=2021-08-17T21:57:55.768341371Z caller=grpc_logging.go:38 method=/logproto.Pusher/Push duration=1.39409ms err="rpc error: code = Code(400) desc = entry with timestamp 2021-08-17 21:56:54.73 +0000 UTC ignored, reason: 'entry out of order' for stream: {SourceContext=\"Rebus.Retry.ErrorTracking.InMemErrorTracker\", app=\"myapp\", container=\"myapp\", errorNumber=\"1\", messageId=\"52xxb042-xxaa-xxxx-813a-31fd6810bb25\", namespace=\"default\", node_name=\"aks-agentpool\", pod=\"myapp-7bd5d4fbcb-gjvxm\"},\ntotal ignored: 1 out of 4" msg="gRPC\n"
`

After this error, the sink does not recover and stops sending logs to Loki.

A similar problem (entry out of order) is discussed [here](https://github.com/grafana/loki/issues/3379))

We are using  version 6.0.1

### PR DESCRIPTION
**Pull request by raphaelquati**
In some situations, multithread applications can write logs in an "out of order" manner.

The sink object is a singleton and the source queue is synchronized, so it's safe to use the current timestamp at [ILogEventSink.Emit](https://github.com/serilog-contrib/serilog-sinks-grafana-loki/blob/53aa24306a197028ab8768ddce5da5004fe815b4/src/Serilog.Sinks.Grafana.Loki/LokiSink.cs#L62)

Resolves #49

### CODE REVIEW

File: src/Serilog.Sinks.Grafana.Loki/LokiBatchFormatter.cs [hunk 1]
```diff
@@ -125,32 +131,39 @@ public void Format(IReadOnlyCollection<LogEvent> logEvents, ITextFormatter forma
...
 
-        private static void GenerateEntry(LogEvent logEvent, ITextFormatter formatter, LokiStream stream, IEnumerable<string> labels)
+        private void GenerateEntry((LogEvent Event, DateTimeOffset Timestamp) logEventTuple, ITextFormatter formatter, LokiStream stream, IEnumerable<string> labels)
         {
             var buffer = new StringWriter(new StringBuilder(DefaultWriteBufferCapacity));
 
+            DateTimeOffset timestamp = logEventTuple.Event.Timestamp;
+            if (_useInternalTimestamp)
+            {
+                logEventTuple.Event.AddPropertyIfAbsent(new LogEventProperty("Time", new ScalarValue(logEventTuple.Event.Timestamp)));
```
Comment (mishamyte) [inline src/Serilog.Sinks.Grafana.Loki/LokiBatchFormatter.cs:141]: Could we call this event `Timestamp`?

**Review by mishamyte — COMMENTED** [review]
Could you cover this behavior with some unit tests, please?

File: src/Serilog.Sinks.Grafana.Loki/LokiBatchFormatter.cs (hunk 1 above)
Comment (raphaelquati) [reply src/Serilog.Sinks.Grafana.Loki/LokiBatchFormatter.cs:141]: > Could you cover this behavior with some unit tests, please?
Sure!

Timestamp is a read-only property of [LogEvent](https://github.com/serilog/serilog/blob/b6bb807422c746413d029968567964e6f0515c4b/src/Serilog/Events/LogEvent.cs#L76)

Comment (raphaelquati) [reply src/Serilog.Sinks.Grafana.Loki/LokiBatchFormatter.cs:141]: I'll rename the member "Event"

Comment (mishamyte) [reply src/Serilog.Sinks.Grafana.Loki/LokiBatchFormatter.cs:141]: Hi @raphaelquati!

Any updated to your pull request?

Comment (raphaelquati) [reply src/Serilog.Sinks.Grafana.Loki/LokiBatchFormatter.cs:141]: Hi.... sorry for my late response (holiday in Brazil) :)
I'll try to update the code today.

Comment (mishamyte) [reply src/Serilog.Sinks.Grafana.Loki/LokiBatchFormatter.cs:141]: No worries)

Hope you are having a great holidays)

**Comment by mishamyte** [comment]
Seems like some tests are failing

**Pull request merged**